insecure/adversary/attacker.go

```diff
@@ -0,0 +1,148 @@
+package adversary
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"sync"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/rs/zerolog"
+	"google.golang.org/grpc"
+
+	"github.com/onflow/flow-go/insecure"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/component"
+	"github.com/onflow/flow-go/module/irrecoverable"
+	"github.com/onflow/flow-go/network"
+)
+
+const networkingProtocolTCP = "tcp"
+
+// Attacker implements the adversarial domain that is orchestrating an attack through corrupted nodes.
+type Attacker struct {
+	component.Component
+	address      string
+	server       *grpc.Server
+	logger       zerolog.Logger
+	orchestrator insecure.AttackOrchestrator
+	codec        network.Codec
+	cm           *component.ComponentManager
+}
+
+func NewAttacker(logger zerolog.Logger, address string, codec network.Codec, orchestrator insecure.AttackOrchestrator) (*Attacker, error) {
+	attacker := &Attacker{
+		orchestrator: orchestrator,
+		logger:       logger,
+		codec:        codec,
+		address:      address,
+	}
+
+	// setting lifecycle management module.
+	cm := component.NewComponentManagerBuilder().
+		AddWorker(func(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+			// starts up gRPC server of attacker at given address.
+			s := grpc.NewServer()
+			insecure.RegisterAttackerServer(s, attacker)
+			ln, err := net.Listen(networkingProtocolTCP, attacker.address)
+			if err != nil {
+				ctx.Throw(fmt.Errorf("could not listen on specified address: %w", err))
+			}
+			attacker.server = s
+
+			wg := sync.WaitGroup{}
+			wg.Add(1)
+			go func() {
+				wg.Done()
```

Comment:
the `wg.Done()` wouldn't guarantee that the gRPC server starts severing. I had a similar issue and resorted to checking a non-nil port was returned.

Reply:
my bad - now seeing that the listener (and hence the port) is already open by this time.